source/common/quic/envoy_quic_server_preferred_address_config_factory.h

```diff
@@ -15,14 +15,29 @@ class EnvoyQuicServerPreferredAddressConfig {
 public:
   virtual ~EnvoyQuicServerPreferredAddressConfig() = default;
 
+  // The set of addresses used to configure the server preferred addresses.
+  struct Addresses {
+    // Addresses that client is requested to use.
+    quic::QuicSocketAddress ipv4_;
+    quic::QuicSocketAddress ipv6_;
+
+    // If destination NAT is applied between the client and Envoy, the addresses that
+    // Envoy will see for client traffic to the server preferred address. If this is not
+    // set, Envoy will expect to receive server preferred address traffic on the above addresses.
+    //
+    // A DNAT address will be ignored if the corresponding SPA address is not set.
+    quic::QuicSocketAddress dnat_ipv4_;
+    quic::QuicSocketAddress dnat_ipv6_;
+  };
+
   /**
    * Called during config loading.
    * @param local_address the configured default listening address.
    * Returns a pair of the server preferred addresses in form of {IPv4, IPv6} which will be used for
    * the entire life time of the QUIC listener. An uninitialized address value means no preferred
    * address for that address family.
    */
-  virtual std::pair<quic::QuicSocketAddress, quic::QuicSocketAddress>
+  virtual Addresses
```

Comment:
Why this interface needs to be changed? And why the proto struct is more preferrable?

Reply:
Because it is now returning 4 addresses instead of two. And a `std::tuple` of 4 elements is harder to read than a struct with four elements.